sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -523,22 +523,62 @@ pi_result piMemBufferCreate(pi_context context, pi_mem_flags flags, size_t size,
                             void *host_ptr, pi_mem *ret_mem,
                             const pi_mem_properties *properties) {
   pi_result ret_err = PI_INVALID_OPERATION;
-  clCreateBufferWithPropertiesINTEL_fn FuncPtr = nullptr;
-
-  if (properties)
+  if (properties) {
+    clCreateBufferWithPropertiesINTEL_fn FuncPtr = nullptr;
+    const size_t propSize = sizeof(properties) / sizeof(pi_mem_properties);
     // First we need to look up the function pointer
     ret_err = getExtFuncFromContext<clCreateBufferWithPropertiesName,
                                     clCreateBufferWithPropertiesINTEL_fn>(
         context, &FuncPtr);
+    if (FuncPtr) {
+      std::vector<pi_mem_properties> supported(properties,
+                                               properties + propSize);
+      // Go through buffer properties. If there is one, that shall be propagated
+      // to an OpenCL runtime - check if this property is being supported.
+      for (auto prop = supported.begin(); prop != supported.end(); ++prop) {
+        // Check if PI_MEM_CHANNEL_INTEL property is supported. If it's not -
+        // just ignore it, as it's an optimization hint.
+        if (*prop == PI_MEM_CHANNEL_INTEL) {
+          size_t deviceCount;
+          cl_int ret_err =
+              clGetContextInfo(cast<cl_context>(context), CL_CONTEXT_DEVICES, 0,
+                               nullptr, &deviceCount);
+          if (ret_err != CL_SUCCESS || deviceCount < 1)
+            return PI_INVALID_CONTEXT;
+          std::vector<cl_device_id> devicesInCtx(deviceCount);
+          ret_err = clGetContextInfo(
+              cast<cl_context>(context), CL_CONTEXT_DEVICES,
+              deviceCount * sizeof(cl_device_id), devicesInCtx.data(), nullptr);
+
+          size_t retSize;
+          ret_err = clGetDeviceInfo(devicesInCtx[0], CL_DEVICE_EXTENSIONS, 0,
```

Comment:
Shouldn't you check all the devices in the context? What does this property mean when allocating for multi-device contexts?

Reply:
Also I think we should cache these queries somehow and not repeat on every request to allocate a buffer.

Reply:
An yet, why is it not using the same code as in line 1015?

Reply:
> An yet, why is it not using the same code as in line 1015?

Don't get this comment.

In general this part was re-worked: created a helper function `getSupportedExtensionsWithinContext`, that queries for common extensions for all devices within a context. List of the extensions is being cached - I think it might be useful not only for this feature. For this feature: each time, when we create a buffer - we search through this list to find out if a particular extension is supported and add/delete the appropriate buffer property.